daemon/algod/api/server/v2/dryrun.go

```diff
@@ -0,0 +1,295 @@
+		if ddr.scratchActive != nil {
+			ddr.history[lasti].Scratch = ddr.history[lasti].Scratch[:len(ddr.scratchActive)]
+		} else {
+			ddr.history[lasti].Scratch = nil
+			return
+		}
+	}
+	scratchlen := maxActive + 1
+	if len(ddr.scratchActive) > scratchlen {
+		scratchlen = len(ddr.scratchActive)
+	}
+	ddr.history[lasti].Scratch = ddr.history[lasti].Scratch[:scratchlen]
+	for i := range ddr.history[lasti].Scratch {
+		if !ddr.scratchActive[i] {
+			ddr.history[lasti].Scratch[i].Type = ""
+		}
+	}
+}
+
+// Register is fired on program creation (DebuggerHook interface)
+func (ddr *dryrunDebugReceiver) Register(state *logic.DebugState) error {
+	ddr.history = append(ddr.history, *state)
+	ddr.updateScratch()
+	return nil
+}
+
+// Update is fired on every step (DebuggerHook interface)
+func (ddr *dryrunDebugReceiver) Update(state *logic.DebugState) error {
+	// see go-algorand/data/transactions/logic/debugger.go refreshDebugState() for which fields are updated
+	ds := logic.DebugState{
+		PC:      state.PC,
+		Line:    state.Line,
+		Error:   state.Error,
+		Stack:   state.Stack,
+		Scratch: state.Scratch,
+	}
+	ddr.history = append(ddr.history, ds)
+	ddr.updateScratch()
+	return nil
+}
+
+// Complete is called when the program exits (DebuggerHook interface)
+func (ddr *dryrunDebugReceiver) Complete(state *logic.DebugState) error {
+	return ddr.Update(state)
+}
+
+// LedgerForLogic
+type dryrunLedger struct {
+	dr *DryrunRequest
+}
+
+// LedgerForLogic
+func (dl *dryrunLedger) Balance(addr basics.Address) (basics.MicroAlgos, error) {
+	return basics.MicroAlgos{Raw: 0}, nil
+}
+func (dl *dryrunLedger) Round() basics.Round {
+	return basics.Round(dl.dr.Round)
+}
+func (dl *dryrunLedger) AppGlobalState(appIdx basics.AppIndex) (basics.TealKeyValue, error) {
+	for _, app := range dl.dr.Apps {
+		if app.AppIndex == uint64(appIdx) {
+			return app.Params.GlobalState, nil
+		}
+	}
+	return nil, nil
+}
+func (dl *dryrunLedger) AppLocalState(addr basics.Address, appIdx basics.AppIndex) (basics.TealKeyValue, error) {
+	for _, st := range dl.dr.AccountAppStates {
+		if appIdx == st.AppIndex && addr == st.Account {
+			return st.State.KeyValue, nil
+		}
+	}
+	return nil, nil
+}
+func (dl *dryrunLedger) AssetHolding(addr basics.Address, assetIdx basics.AssetIndex) (basics.AssetHolding, error) {
+	for _, acct := range dl.dr.Accounts {
+		// TODO: check all address decodes when we first receive the dryrun request
+		xaddr, err := basics.UnmarshalChecksumAddress(acct.Address)
+		if err != nil {
+			continue
+		}
+		if xaddr == addr {
+			for _, ah := range *acct.Assets {
+				if ah.AssetId == uint64(assetIdx) {
+					return basics.AssetHolding{Amount: ah.Amount, Frozen: ah.IsFrozen}, nil
+				}
+			}
+		}
+	}
+	return basics.AssetHolding{Amount: 0, Frozen: false}, nil
+}
+func (dl *dryrunLedger) AssetParams(addr basics.Address, assetIdx basics.AssetIndex) (basics.AssetParams, error) {
+	// TODO? maybe not needed for dryrun
+	return basics.AssetParams{}, nil
```

Comment:
I think it would be nice if we supported this -- it's the only missing method from this endpoint otherwise.

Reply:
I'm tempted to write a bunch of unit tests and panic here and prove it isn't called, I think nothing dryrun does cares about this detail of ledger asset handling?